infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -141,87 +150,31 @@ private void connect(ClusterDescriptor remoteCluster, ReplicationSession session
-                                  Set<String> intersection) {
-
-        long oldTopologyConfigId = replicationContext.getTopologyConfigId();
-        replicationContext.setTopologyConfigId(newConfig.getTopologyConfigId());
-
-        // Get all subscribers
-        Set<ReplicationSubscriber> subscribers = replicationContext.getConfig().getReplicationSubscriberToStreamsMap().keySet();
-
-        // Remove Sinks that are not in the new config
-        for (String clusterId : sinksToRemove) {
-            for (ReplicationSubscriber subscriber : subscribers) {
-                stopLogReplicationRuntime(new ReplicationSession(clusterId, subscriber));
-            }
-        }
-
-        // Start the newly added Sinks
-        for (String clusterId : sinksToAdd) {
-            ClusterDescriptor clusterInfo = newConfig.getSinkClusters().get(clusterId);
-            for (ReplicationSubscriber subscriber : subscribers) {
-                startLogReplicationRuntime(clusterInfo, new ReplicationSession(clusterId, subscriber));
-            }
-        }
-
+    public void refreshRuntime(LogReplicationSession session, ClusterDescriptor cluster, long topologyConfigId) {
         // The connection id or other transportation plugin's info could've changed for existing Sink clusters,
         // updating the routers will re-establish the connection to the correct endpoints/nodes
-        for (String clusterId : intersection) {
-            ClusterDescriptor clusterInfo = newConfig.getSinkClusters().get(clusterId);
-            for (ReplicationSubscriber subscriber : subscribers) {
-                runtimeToRemoteSession.get(new ReplicationSession(clusterId, subscriber))
-                    .updateRouterClusterDescriptor(clusterInfo);
-            }
-        }
-
-        if (oldTopologyConfigId != newConfig.getTopologyConfigId()) {
-            updateRuntimeConfigId(newConfig);
-        }
+        sessionRuntimeMap.get(session).updateRouterClusterDescriptor(cluster, topologyConfigId);
     }
 
     /**
-     * Stop the current log replication event and start a full snapshot sync for the given remote cluster.
+     * Stop the current log replication event and start a full snapshot sync for the given session.
      */
     public void enforceSnapshotSync(DiscoveryServiceEvent event) {
-        CorfuLogReplicationRuntime sinkRuntime = runtimeToRemoteSession.get(
-            ReplicationSession.getDefaultReplicationSessionForCluster(event.getRemoteClusterInfo().getClusterId()));
-        if (sinkRuntime == null) {
-            log.warn("Failed to start enforceSnapshotSync for cluster {} as no runtime to it was found",
-                event.getRemoteClusterInfo().getClusterId());
+        CorfuLogReplicationRuntime runtime = sessionRuntimeMap.get(event.getSession());
+        if (runtime == null) {
+            log.warn("Failed to enforce snapshot sync for session {}",
+                event.getSession());
```

Comment:
A question to clarify, this forced snapshot sync workflow is for clients' `forceFullSync` usage right? as I saw it went through eventTable -> stream listener -> discoveryService -> fsm to get triggerred. 

Then for our internal usage, like in the case a destination is added to a new group, I thought we can reach out to the session's fsm directly right (without going through the eventTable)? Would it be good to have an overloaded method for it?
```
public void enforceSnapshotSync(LogReplicationSession session)
```
I could add it in my next PR if this approach is correct